### DISCUSSION TITLE
Undefined platform

### DISCUSSION THREAD

**Discussion by Tampa** [jart/cosmopolitan#1329] [Q&A] [answered]
Trying following code to determine the platform an application is run on to handle different paths to libraries:

```
#if defined(_WIN32)
    #define PLATFORM_NAME "windows" // Windows
#elif defined(_WIN64)
    #define PLATFORM_NAME "windows" // Windows
#elif defined(__linux__)
    #define PLATFORM_NAME "linux" // Debian, Ubuntu, Gentoo, Fedora, openSUSE, RedHat, Centos and other
#elif defined(__APPLE__) && defined(__MACH__)
    #include <TargetConditionals.h>
    #if TARGET_OS_MAC == 1
        #define PLATFORM_NAME "osx" // Apple OSX
    #endif
#else
	#define PLATFORM_NAME "error" // error
#endif

int main(int argc, char *argv[]) {
    fprintf(stdout, "Determined platform is %s\n", PLATFORM_NAME);
}
```

However, this always returns the determined platform as error. In gcc this compiles even without the else branch and properly determines the platform. Is this not supported or what am I missing?

**Comment by GabrielRavier** [comment, marked as answer]
Those are checks you need to make at runtime with functions like `IsWindows`/`IsLinux`/`IsXnu`/`IsBsd`/`IsAarch64`/`IsFreebsd`/`IsOpenbsd`/`IsNetBsd`/`IsMetal`/etc.

(making such checks at compile time would not be possible by definition since the platform Cosmopolitan runs on is determined at runtime)

**Reply by Tampa** [reply]
Good to know, thanks!

In the meantime I removed them to move on and see if I could get past the issue I had with gcc, but while I got a bit further, unfortunately new problems.

I'm trying to write a simple wrapper to launch a .Net application via system() call to "dotnet Program.dll" which works fine on Linux and on Windows through the terminal and command line, but trying to launch the Wrapper.exe via double click I get a number of issues.

Either it tells me the dotnet file or directory doesn't exist, evidently not recognizing it as a command, or when using the absolute path to dotnet.exe it just refuses to start: "The application was unable to start correctly 0xc0000142", which appears on any application I attempt to start this way. At least, unlike gcc, it at least opens a command line on Windows.

Do I have to import the environment in some way for it to find the dotnet command?

**Reply by jart** [reply]
Try building the following .c program:

```c
#include <stdlib.h>

int main(int argc, char *argv[]) {
  system("notepad.exe");
}
```

If you run that, then does it cause the notepad program to open?

**Reply by Tampa** [reply]
Yes that works fine.

I tried this next:
```
#include <stdlib.h>
#include <stdio.h>

int main(int argc, char *argv[]) {
  system("dotnet --version");
  while (getchar() != '\n');
}
```
dotnet: No such file or directory

`system("\"C:\\Program Files\\dotnet\\dotnet.exe\" --version");` Instead produced the 0xc0000142 error.

I then just put dotnet.exe in there and placed the executable right next to that into the folder, same 0xc error. Opening a cmd in there and doing dotnet.exe --version worked fine though.

Then tried `cmd.exe /c \"dotnet.exe --version\"` and got the 0xc error on the cmd.exe itself.

Just `cmd /c ...` I get cmd: No such file...

This is really bizarre behavior.

**Reply by Tampa** [reply]
```
#include <stdlib.h>

int main(int argc, char *argv[]) {
  system("dotnet.exe --version");
}
```

Placing the executable into the dotnet folder and running in cmd as "Test.exe" I get a version return, trying to then do the same by double clicking on Test.exe I receive the 0xc error instead. Somehow it really doesn't like to be double clicked on to launch.

**Reply by Tampa** [reply]
I'm stumped. I cannot seem to really find anything why that error shows up in this case. Most are related to issues in setup, but that should be handled by APE.

The strange part is that it does open a cmd and starts writing in there using printf as well, but when it comes to actually running dotnet it throws up.

If I run it from a cmd to begin with it works fine, though I suppose it also doesn't open another cmd as I would expect it might.
